### ISSUE TITLE
ex 2 level 3

### ISSUE THREAD

**Issue by kelleyhan25** [nus-cs2030/2223-s2#145]
Hi all! Was wondering if anyone had any tips for level 3 exercise 2. The wording of the question is confusing me and I'm not sure where to start. Thank you in advance!

**Comment by hhuilingg** [comment]
hi! you may also want to look at #97

**Comment by kelleyhan25** [comment]
> There are two queues: Active and Expired.
> 
> When a cruise arrives at time X, you will look through services the Active queue. If a service in the Active queue has already finished at time X, you will remove it from the Active queue and put it in the Expired queue.
> 
> Next, you will look through the Expired queue. Get the loaders from the expired services and assign them to the cruise by creating a new service. Add this new service to the Active queue. Repeat until the cruise has sufficient loaders.
> 
> If there were not enough loaders from the Expired queue (which will be the case for the first cruise that arrives), create new loaders and assign them to the cruise by creating a new service. Likewise, add these to the Active queue.
> 
> Now another cruise arrives at time X+1, and the cycle above repeats.

Thank you!! So does this mean that when a cruise arrives, I have to create a new Service for it? And then put that service into the active/expired queue?

**Comment by trwstin** [comment]
> > There are two queues: Active and Expired.
> > When a cruise arrives at time X, you will look through services the Active queue. If a service in the Active queue has already finished at time X, you will remove it from the Active queue and put it in the Expired queue.
> > Next, you will look through the Expired queue. Get the loaders from the expired services and assign them to the cruise by creating a new service. Add this new service to the Active queue. Repeat until the cruise has sufficient loaders.
> > If there were not enough loaders from the Expired queue (which will be the case for the first cruise that arrives), create new loaders and assign them to the cruise by creating a new service. Likewise, add these to the Active queue.
> > Now another cruise arrives at time X+1, and the cycle above repeats.
> 
> Thank you!! So does this mean that when a cruise arrives, I have to create a new Service for it? And then put that service into the active/expired queue?

Yes, the new Service should go into the Active queue. It will only go to the Expired queue when it expires.

**Comment by kelleyhan25** [comment]
> > > There are two queues: Active and Expired.
> > > When a cruise arrives at time X, you will look through services the Active queue. If a service in the Active queue has already finished at time X, you will remove it from the Active queue and put it in the Expired queue.
> > > Next, you will look through the Expired queue. Get the loaders from the expired services and assign them to the cruise by creating a new service. Add this new service to the Active queue. Repeat until the cruise has sufficient loaders.
> > > If there were not enough loaders from the Expired queue (which will be the case for the first cruise that arrives), create new loaders and assign them to the cruise by creating a new service. Likewise, add these to the Active queue.
> > > Now another cruise arrives at time X+1, and the cycle above repeats.
> > 
> > 
> > Thank you!! So does this mean that when a cruise arrives, I have to create a new Service for it? And then put that service into the active/expired queue?
> 
> Yes, the new Service should go into the Active queue. It will only go to the Expired queue when it expires.

Thanks!

**Comment by kelleyhan25** [comment]
> > > There are two queues: Active and Expired.
> > > When a cruise arrives at time X, you will look through services the Active queue. If a service in the Active queue has already finished at time X, you will remove it from the Active queue and put it in the Expired queue.
> > > Next, you will look through the Expired queue. Get the loaders from the expired services and assign them to the cruise by creating a new service. Add this new service to the Active queue. Repeat until the cruise has sufficient loaders.
> > > If there were not enough loaders from the Expired queue (which will be the case for the first cruise that arrives), create new loaders and assign them to the cruise by creating a new service. Likewise, add these to the Active queue.
> > > Now another cruise arrives at time X+1, and the cycle above repeats.
> > 
> > 
> > Thank you!! So does this mean that when a cruise arrives, I have to create a new Service for it? And then put that service into the active/expired queue?
> 
> Yes, the new Service should go into the Active queue. It will only go to the Expired queue when it expires.

Sorry, quick clarification question - wording is throwing me off a bit. Are we supposed to import the queue ADT and use that, or use an ImList? Thanks again!

**Comment by trwstin** [comment]
Just 

> > > > There are two queues: Active and Expired.
> > > > When a cruise arrives at time X, you will look through services the Active queue. If a service in the Active queue has already finished at time X, you will remove it from the Active queue and put it in the Expired queue.
> > > > Next, you will look through the Expired queue. Get the loaders from the expired services and assign them to the cruise by creating a new service. Add this new service to the Active queue. Repeat until the cruise has sufficient loaders.
> > > > If there were not enough loaders from the Expired queue (which will be the case for the first cruise that arrives), create new loaders and assign them to the cruise by creating a new service. Likewise, add these to the Active queue.
> > > > Now another cruise arrives at time X+1, and the cycle above repeats.
> > > 
> > > 
> > > Thank you!! So does this mean that when a cruise arrives, I have to create a new Service for it? And then put that service into the active/expired queue?
> > 
> > 
> > Yes, the new Service should go into the Active queue. It will only go to the Expired queue when it expires.
> 
> Sorry, quick clarification question - wording is throwing me off a bit. Are we supposed to import the queue ADT and use that, or use an ImList? Thanks again!

ImList will do

**Comment by choongzhanhong** [comment]
I feel a little silly having trouble over this problem, but [this](https://stackoverflow.com/a/60859469) provided such an elegant solution to the problem of removing elements while iterating.
